constructor/header.sh

```diff
@@ -1,89 +1,125 @@
-#!/bin/bash
+#!/bin/sh
 #
 # NAME:  __NAME__
 # VER:   __VERSION__
 # PLAT:  __PLAT__
 # BYTES: @SIZE_BYTES@
 # LINES: @LINES@
 # MD5:   __MD5__
-
 #if osx
 unset DYLD_LIBRARY_PATH
 #else
 export OLD_LD_LIBRARY_PATH=$LD_LIBRARY_PATH
 unset LD_LIBRARY_PATH
 #endif
 
-echo "$0" | grep '\.sh$' >/dev/null
-if (( $? )); then
+if ! echo "$0" | grep '\.sh$' >/dev/null; then
     echo 'Please run using "bash" or "sh", but not "." or "source"' >&2
     return 1
 fi
 
-THIS_DIR=$(cd $(dirname $0); pwd)
-THIS_FILE=$(basename $0)
+# Determine RUNNING_SHELL; if SHELL is non-zero use that.
+if [ -n "$SHELL" ]; then
+    RUNNING_SHELL="$SHELL"
+else
+    if [ "$(uname)" = "Darwin" ]; then
+        RUNNING_SHELL=/bin/bash
+    else
+        RUNNING_SHELL=$(ps -p $$ -oargs=)
+    fi
+fi
+if [ -z "$RUNNING_SHELL" ] || [ ! -f "$RUNNING_SHELL" ]; then
+    if [ -f /bin/bash ]; then
+        RUNNING_SHELL=/bin/bash
+    else
+        if [ -f /bin/sh ]; then
+            RUNNING_SHELL=/bin/sh
+        fi
+    fi
+fi
+
+THIS_DIR=$(DIRNAME=$(dirname "$0"); cd "$DIRNAME"; pwd)
+THIS_FILE=$(basename "$0")
 THIS_PATH="$THIS_DIR/$THIS_FILE"
 PREFIX=__DEFAULT_PREFIX__
 BATCH=0
 FORCE=0
 SKIP_SCRIPTS=0
-
-while getopts "bfhp:su" x; do
-    case "$x" in
-        h)
-            echo "usage: $0 [options]
+USAGE="
+usage: $0 [options]
 
 Installs __NAME__ __VERSION__
 
-    -b           run install in batch mode (without manual intervention),
-                 it is expected the license terms are agreed upon
-    -f           no error if install prefix already exists
-    -h           print this help message and exit
-    -p PREFIX    install prefix, defaults to $PREFIX
-    -s           skip running pre/post-link/install scripts
-    -u           update an existing installation
+-b           run install in batch mode (without manual intervention),
+     it is expected the license terms are agreed upon
+-f           no error if install prefix already exists
+-h           print this help message and exit
+-p PREFIX    install prefix, defaults to $PREFIX, must not contain spaces.
+-s           skip running pre/post-link/install scripts
+-u           update an existing installation
 "
+
+OPTS=$(getopt bfhp:su "$*" 2>/dev/null)
+if [ ! $? ]; then
+    echo "$USAGE"
+    exit 2
+fi
+
+eval set -- "$OPTS"
+
+while true; do
+    case "$1" in
```

Comment:
This isn't actually trying to parse flags properly, is it?  I use `-bfp` all the time, and have so publicly in a lot of issues.  With the parsing now, wouldn't that have to change to `-b -f -p`?  Should we use `getopts` here instead?